tests/new_api_test.cpp

```diff
@@ -1345,6 +1353,270 @@ void test_remove_corrupted(const ioremap::elliptics::newapi::session &session) {
+		512,
+		std::string{"key data"},
+		1024,
+		(session.get_ioflags() & DNET_IO_FLAGS_CACHE) != 0
+	};
+
+	std::map<dnet_id, tests::record> records;
+	std::set<dnet_id> responses;
+
+	auto check_all_responses_presence = [&records, &responses] () {
+		for (const auto &pair : records) {
+			auto &key = pair.first;
+			BOOST_REQUIRE(responses.find(key) != responses.end());
+		}
+	};
+
+	std::vector<int> groups{1, 2, 3, 4, 5, 6};
+
+	/*
+	 * Step 1. Prepare test data: write keys to multiple groups. Each key will be presented
+	 * exactly once in every group.
+	 */
+	const static size_t NUM_KEYS_IN_GROUP = 100;
+	const static size_t NUM_KEYS_TOTAL = NUM_KEYS_IN_GROUP * groups.size();
+
+	std::vector<std::tuple<ioremap::elliptics::newapi::async_write_result,
+			       decltype(records)::const_iterator>> write_results;
+	write_results.reserve(NUM_KEYS_TOTAL);
+
+	std::vector<dnet_id> ids;
+	ids.reserve(NUM_KEYS_TOTAL);
+
+	auto s = session.clone();
+	s.set_filter(ioremap::elliptics::filters::all_with_ack);
+	s.set_trace_id(rand());
+	s.set_user_flags(record.user_flags);
+	s.set_timestamp(record.timestamp);
+	s.set_json_timestamp(record.json_timestamp);
+
+	for (size_t i = 0; i < NUM_KEYS_IN_GROUP; ++i) {
+		for (const int group_id : groups) {
+			record.key = key("bulk_key_" + std::to_string(i));
+			record.key.transform(s);
+			record.key.set_group_id(group_id);
+
+			auto unique_suffix = std::to_string(group_id * NUM_KEYS_IN_GROUP + i);
+			record.json = "{\"key\": \"bulk_json_" + unique_suffix + "\"}";
+			record.data = "bulk_data_" + unique_suffix;
+
+			s.set_groups({group_id});
+			auto async = s.write(record.key,
+					     record.json, record.json_capacity,
+					     record.data, record.data_capacity);
+
+			auto it = records.emplace(record.key.id(), record).first;
+
+			write_results.emplace_back(std::move(async), it);
+			ids.emplace_back(record.key.id());
+		}
+	}
+
+	for (auto &r : write_results) {
+		auto &async = std::get<0>(r);
+		const auto &record = std::get<1>(r)->second;
+		check_lookup_result(async, DNET_CMD_WRITE_NEW, record, 1);
+	}
+
+	/*
+	 * Step 2. Check bulk read in normal conditions: read keys from all groups
+	 * and check its data. Check it against different read methods:
+	 * bulk_read, bulk_read_data, bulk_read_json.
+	 */
+	size_t count = 0;
+	auto check_read_result = [&] (ioremap::elliptics::newapi::async_read_result &async,
+				      bool check_json, bool check_data) {
+		responses.clear();
+		size_t count = 0;
+		for (const auto &result: async) {
+			BOOST_REQUIRE_EQUAL(result.status(), 0);
+			BOOST_REQUIRE_EQUAL(result.command()->cmd, DNET_CMD_BULK_READ_NEW);
+			BOOST_REQUIRE_EQUAL(result.error().code(), 0);
+			BOOST_REQUIRE_EQUAL(result.error().message(), "");
+
+			const auto cmd = result.command();
+			responses.emplace(cmd->id);
+
+			auto it = records.find(cmd->id);
+			BOOST_REQUIRE(it != records.end());
+			auto ref = it->second;
+
+			auto record_info = result.record_info();
+
+			if (check_json) {
```

Comment:
In any case `record_info.json_*` and `record_info.data_*` should be filled with valid values, so they should be checked. If `check_json` is false, test should check that no json has been read. If `check_data` is false, test should check that no data has been read.